dds/DCPS/XTypes/TypeLookupService.cpp

```diff
@@ -123,6 +125,164 @@ void TypeLookupService::add_type_objects_to_cache(const TypeIdentifier& ti, cons
   }
 }
 
+void TypeLookupService::add_type_identifiers_to_cache(const TypeIdentifierPairSeq& pair)
+{
+  ACE_GUARD(ACE_Thread_Mutex, g, mutex_);
+  for (ACE_UINT32 i = 0; i < pair.length(); ++i) {
+    //type_identifier1 is assumed to be the complete type identifier
+    const TypeIdentifierMap::iterator pos = type_identifier_map_.find(pair[i].type_identifier1);
+    if (pos == type_identifier_map_.end()) {
+      type_identifier_map_.insert(std::make_pair(pair[i].type_identifier1, pair[i].type_identifier2));
+    }
+  }
+}
+
+bool TypeLookupService::get_minimal_type_identifier(const TypeIdentifier& cti, TypeIdentifier& mti)
+{
+  ACE_GUARD_RETURN(ACE_Thread_Mutex, g, mutex_, false);
+  const TypeIdentifierMap::const_iterator it = type_identifier_map_.find(cti);
+  if (it != type_identifier_map_.end()) {
+    mti = it->second;
+    return true;
+  }
+  ACE_ERROR((LM_ERROR, ACE_TEXT("get_minimal_type_identifier(): Unable to convert complete type identifier to minimal\n")));
+  return false;
+}
+
+bool TypeLookupService::complete_to_minimal_struct(const CompleteStructType& ct, MinimalStructType& mt) {
+  mt.struct_flags = ct.struct_flags;
+  if (!get_minimal_type_identifier(ct.header.base_type, mt.header.base_type)) {
+    ACE_ERROR((LM_ERROR, ACE_TEXT("complete_to_minimal_union(): Failed to get minimal type identifier\n")));
+    return false;
+  }
+  mt.member_seq.length(ct.member_seq.length());
+  for(unsigned long i = 0; i < ct.member_seq.length(); ++i) {
+    mt.member_seq[i].common = ct.member_seq[i].common;
+    hash_member_name(mt.member_seq[i].detail.name_hash, ct.member_seq[i].detail.name);
+  }
+  return true;
+}
+
+ bool TypeLookupService::complete_to_minimal_union(const CompleteUnionType& ct, MinimalUnionType& mt) {
+  mt.union_flags = ct.union_flags;
+  mt.discriminator.common = ct.discriminator.common;
+  if (!get_minimal_type_identifier(ct.discriminator.common.type_id, mt.discriminator.common.type_id)) {
+    ACE_ERROR((LM_ERROR, ACE_TEXT("complete_to_minimal_union(): Failed to get minimal type identifier\n")));
+    return false;
+  }
+  mt.member_seq.length(ct.member_seq.length());
+  for(unsigned long i = 0; i < ct.member_seq.length(); ++i) {
+    mt.member_seq[i].common = ct.member_seq[i].common;
```

Comment:
Similar comment to the conversion for struct above. This applies to the other conversion methods for the other types too.